pkg/controller/awselb/aws_elb_controller_test.go

```diff
@@ -0,0 +1,298 @@
+			lastSuccessfulSync:           time.Now().Add(-5 * time.Second),
+			generation:                   int64(1),
+			lastSuccessfulSyncGeneration: int64(1),
+			resyncExpected:               false,
+		},
+		{
+			name:                         "re-sync due",
+			lastSuccessfulSync:           time.Now().Add(-9 * time.Hour),
+			generation:                   int64(1),
+			lastSuccessfulSyncGeneration: int64(1),
+			resyncExpected:               true,
+		},
+		{
+			name:                         "re-sync for generation change",
+			lastSuccessfulSync:           time.Now().Add(-5 * time.Second),
+			generation:                   int64(4),
+			lastSuccessfulSyncGeneration: int64(1),
+			resyncExpected:               true,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
+
+			kubeClient := &clientgofake.Clientset{}
+			clustopClient := &clustopclientfake.Clientset{}
+			capiClient := &capiclientfake.Clientset{}
+			capiInformers := capiinformers.NewSharedInformerFactory(capiClient, 0)
+
+			status := clustopv1.AWSMachineProviderStatus{}
+			status.InstanceID = aws.String("preserveme")
+			if !tc.lastSuccessfulSync.IsZero() {
+				status.LastELBSync = &metav1.Time{Time: tc.lastSuccessfulSync}
+			}
+			status.LastELBSyncGeneration = tc.lastSuccessfulSyncGeneration
+			cluster := testCluster(t)
+			machine := testMachine(tc.generation, testMachineName, cluster.Name, clustopv1.NodeTypeMaster, true, &status)
+
+			mockAWSClient := mockaws.NewMockClient(mockCtrl)
+			if tc.resyncExpected {
+				mockTestInstance(mockAWSClient, "i1", machine.Name, testClusterID)
+				mockRegisterInstancesWithELB(mockAWSClient, "i1", controller.ELBMasterExternalName(testClusterID))
+				mockRegisterInstancesWithELB(mockAWSClient, "i1", controller.ELBMasterInternalName(testClusterID))
+			}
+
+			ctrlr := NewController(capiInformers.Cluster().V1alpha1().Machines(),
+				kubeClient, clustopClient, capiClient)
+			ctrlr.clientBuilder = func(kubeClient kubernetes.Interface, mSpec *clustopv1.MachineSetSpec, namespace, region string) (clustopaws.Client, error) {
+				return mockAWSClient, nil
+			}
+
+			err := ctrlr.processMachine(machine)
+			assert.NoError(t, err)
+
+			if tc.resyncExpected {
+				if !assert.Equal(t, 1, len(capiClient.Actions())) {
+					return
+				}
+				action := capiClient.Actions()[0]
+				assert.Equal(t, "update", action.GetVerb())
+
+				updateAction, ok := action.(clientgotesting.UpdateAction)
+				assert.True(t, ok)
+
+				updatedObject := updateAction.GetObject()
+				machine, ok := updatedObject.(*capiv1.Machine)
+				clustopStatus, err := controller.AWSMachineProviderStatusFromClusterAPIMachine(machine)
+				assert.NoError(t, err)
+				assert.NotNil(t, clustopStatus.LastELBSync)
+				assert.True(t, clustopStatus.LastELBSync.Time.After(tc.lastSuccessfulSync))
+				assert.NotEqual(t, int64(0), clustopStatus.LastELBSyncGeneration)
```

Comment:
Test that this equals the generation of the Machine instead of just that it is not zero.